### DISCUSSION TITLE
Chaining 2 commands / assertion in one snapshot?

### DISCUSSION THREAD

**Discussion by paxcodes** [cypress-io/cypress#17296] [Questions and Help] [answered]
The app shows a spinner when a user clicks a certain button. Behind the scenes, 

- the app is making a network request
- shows a spinner until the request has completed
- does not display a final message  until the request has completed

To test, I have this:

```js
    // user clicks button
    cy.get("[data-cy=confirmBtn]").click(); 

    // check that loading spinner is visible and that final message is not visible
    cy.get("[data-cy=loadingSpinner]").should("be.visible")
    cy.get("[data-cy=finalMessage]").should("not.be.visible");
```

The last statement is currently failing because by the time Cypress proceeds with the last statement, the "request" has finished and finalMessage is visible.

My ideas to fix this involving intercepting the request that happens once the button is clicked:

- manually control when to send back the response (from [this blog](https://blog.dai.codes/cypress-loading-state-tests/))
    - But in my case, I don't want the request to send back a response. It seems so much test code for something as simple as, "never send back a response"
- Use an arbitrary delay
    - Arbitrary delays smells like flaky test
    - Having to assign the time it would take the test to finish may not be good practice?

It would be nice if I can chain "get" commands within the same "snapshot" but not sure how or if that can be done in Cypress.

What is the canonical way / best practice of fixing this?

**Comment by paxcodes** [comment, marked as answer]
This may be the closest solution I can get to. Based on https://docs.cypress.io/api/commands/intercept#Returning-a-Promise-2 and [the blog post](https://blog.dai.codes/cypress-loading-state-tests/) I referred to.

My test now looks like,

```js
    // Let's intercept the request so we can control when to receive
    // the response.
    // https://docs.cypress.io/api/commands/intercept#Returning-a-Promise-2
    let resolveRequest;
    const waitForTestToFinish = new Promise((resolve) => {
      resolveRequest = resolve;
    })
    cy.intercept('DELETE', '/api/loans/*', (request) => {
      request.continue((_) => {
        return waitForTestToFinish;
      });
    });

    // user clicks button
    cy.get("[data-cy=confirmBtn]").click(); 

    // check that loading spinner is visible and that final message is not visible 
    // while the request is ongoing.
    cy.get("[data-cy=loadingSpinner]").should("be.visible")
      .get("[data-cy=finalMessage]").should("not.be.visible").then(resolveRequest);
```

If anyone has a "less code" / simpler solution, that'd be really great.

**Comment by bahmutov** [comment]
I would test it similar to the loading indicator test described in https://glebbahmutov.com/blog/negative-assertions/

You can also write complex synchronous assertion logic by putting it into a single `should(callback)` function, like

```js
cy.get('.loading').should(($loading) => {
  // single function with synchronous assertions
  expect($loading).to.be.visible
  const doc = $loading[0].ownerDocument
  const todoList = doc.querySelector('.todo-list')
  expect(todoList).to.be.empty
})
// the loading element goes away and the list appears
cy.get('.loading').should('not.be.visible')
cy.get('.todo-list li').should('have.length', 2)
```

**Reply by paxcodes** [reply]
Thanks for this bahmutov. That's helpful advice. It seems though that I still have to intercept the request to make sure it doesn't "complete" before I am able to make the assertion that the spinner is visible, etc.

Otherwise, by the time Cypress executes the synchronous assertions, the request has already been completed and the assertions fail (loading spinner is gone, the final message is visible).
